database/mssql/scripts/dbscripts/PSP_PIMS_S15_01/Alter Up/02_PSP_PIMS_S15_00_Alter.sql

```diff
@@ -0,0 +1,1086 @@
+GO
+EXEC sp_addextendedproperty 
+	@name = N'MS_Description', @value = N'The data source column value' , 
+	@level0type = N'Schema', @level0name = N'etl', 
+	@level1type = N'Table', @level1name = N'ETL_DATA_SOURCE_TRACE', 
+	@level2type = N'Column', @level2name = N'SRC_COLUMN_VALUE'
+GO
+EXEC sp_addextendedproperty 
+	@name = N'MS_Description', @value = N'Persistent table to trace the source of ETL data to assist in data matching.' , 
+	@level0type = N'Schema', @level0name = N'etl', 
+	@level1type = N'Table', @level1name = N'ETL_DATA_SOURCE_TRACE'
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Create procedure dbo.pims_error_handling
+PRINT N'Create procedure dbo.pims_error_handling'
+GO
+-- ............................................................................................
+
+
+-- --------------------------------------------------------------------------------------------
+-- Create procedure pims_error_handling 
+-- --------------------------------------------------------------------------------------------
+CREATE PROCEDURE [dbo].[pims_error_handling] AS 
+  begin
+    DECLARE @errmsg   nvarchar(2048),
+      @severity tinyint,
+      @state    tinyint,
+      @errno    int,
+      @proc     sysname,
+      @lineno   int
+           
+    SELECT @errmsg = error_message(), @severity = error_severity(),
+      @state  = error_state(), @errno = error_number(),
+      @proc   = error_procedure(), @lineno = error_line()
+
+    IF @errmsg NOT LIKE '***%'
+      BEGIN
+        SELECT @errmsg = '*** ' + coalesce(quotename(@proc), '<dynamic SQL>') + 
+          ', Line ' + ltrim(str(@lineno)) + '. Errno ' + 
+          ltrim(str(@errno)) + ': ' + @errmsg
+      END
+
+    RAISERROR('%s', @severity, @state, @errmsg)
+  end
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Alter table dbo.PIMS_LEASE
+PRINT N'Alter table dbo.PIMS_LEASE'
+GO
+ALTER TABLE [dbo].[PIMS_LEASE]
+	ADD [RETURN_NOTES] nvarchar(4000) NULL
+GO
+EXEC sp_addextendedproperty 
+	@name = N'MS_Description', @value = N'Notes accompanying lease' , 
+	@level0type = N'Schema', @level0name = N'dbo', 
+	@level1type = N'Table', @level1name = N'PIMS_LEASE', 
+	@level2type = N'Column', @level2name = N'RETURN_NOTES'
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Alter table dbo.PIMS_SECURITY_DEPOSIT_RETURN
+PRINT N'Alter table dbo.PIMS_SECURITY_DEPOSIT_RETURN'
+GO
+ALTER TABLE [dbo].[PIMS_SECURITY_DEPOSIT_RETURN] ALTER COLUMN [PAYEE_ADDRESS] nvarchar(500) NULL
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+ALTER TABLE [dbo].[PIMS_SECURITY_DEPOSIT_RETURN]
+	DROP COLUMN [TERMINATION_NOTE]
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+-- Alter table dbo.PIMS_SECURITY_DEPOSIT
+PRINT N'Alter table dbo.PIMS_SECURITY_DEPOSIT'
+GO
+EXEC dbo.sp_rename N'dbo.PIMS_SECURITY_DEPOSIT.TOTAL_AMOUNT' , N'OTHER_DEP_HOLDER_TYPE_DESC', 'COLUMN'
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+ALTER TABLE [dbo].[PIMS_SECURITY_DEPOSIT] ALTER COLUMN [OTHER_DEP_HOLDER_TYPE_DESC] nvarchar(100) NULL
+GO
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
```

Comment:
This table modification will only work (I think) if there is not pre-existing data in the table. 
I'm not a SQL expert but would expect the tool to generate something like:

```sql
ALTER TABLE ... DROP COLUMN "X"
ALTER TABLE ... ADD COLUMN "Y"
```

Instead it is reusing the **same** column by renaming it and then changing the data type. Since the original column is for decimal values I think any existing data would cause this script to throw an error.

Again, I'm just highlighting potential issues in the future - I think this time it would be fine since there's no security deposit data at the moment

@devinleighsmith thoughts?

Reply:
Definitely interesting.  I tested the Alter Up and had no problem but at the time of testing the table was empty.

Reply:
As it turns out, decimal (money in this case) is an easy redefinition of the character representation of a DECIMAL value  to VARCHAR.  The DECIMAL (or MONEY) is stored internally as character numbers and the DECIMAL representation is basically a set of behaviours imposed on the DECIMAL value.  At least that's my understanding of it.

This is not the case for FLOAT values but the cure for that is something like (where Phonenumber is defined as FLOAT):
  ALTER TABLE Customer ALTER COLUMN Phonenumber decimal(25, 0);
  ALTER TABLE Customer ALTER COLUMN Phonenumber varchar(25);

Magic!